core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/visualization/wordCloud/WordCloudOpFinalExec.java

```diff
@@ -38,42 +46,79 @@ public void close() {
         termFreqMap = null;
     }
 
-    @Override
-    public String getParam(String query) {
-        return null;
+    public List<Tuple> normalizeWordCloudTuples() {
+        double minValue = Double.MAX_VALUE;
+        double maxValue = Double.MIN_VALUE;
+
+        for (Map.Entry<String, Integer> e : termFreqMap.entrySet()) {
+            int frequency = e.getValue();
+            minValue = Math.min(minValue, frequency);
+            maxValue = Math.max(maxValue, frequency);
+        }
+        // normalize the font size for wordcloud js
+        // https://github.com/timdream/wordcloud2.js/issues/53
+        List<Tuple> termFreqTuples = new ArrayList<>();
+        for (Map.Entry<String, Integer> e : termFreqMap.entrySet()) {
+            termFreqTuples.add(Tuple.newBuilder().add(
+                    resultSchema,
+                    Arrays.asList(
+                            e.getKey(),
+                            (int) ((e.getValue() - minValue) / (maxValue - minValue) *
+                                    (MAX_FONT_SIZE - MIN_FONT_SIZE) + MIN_FONT_SIZE)
+                    )
+            ).build());
+        }
+        return termFreqTuples;
+    }
+
+    public List<Tuple> calculateResults(List<Tuple> normalizedWordCloudTuples) {
+        List<Tuple> retractions = new ArrayList<>(prevWordCloudTuples);
+        List<Tuple> insertions = new ArrayList<>(normalizedWordCloudTuples);
```

Comment:
I believe this is making copies of the `ArrayList`. Is this necessary?

Reply:
this is not making a copy of the underlying data. This is necessary because we are mutating the two lists.